File: bitcoin-stockmem-ts/src/pipeline/step1-extract.ts
/**
 * Step 1: Event Extraction (LLM_ext).
 */

import { GeminiClient } from "../llm/gemini-client";
import { EXTRACT_PROMPT, fillTemplate } from "../llm/prompts";
import { parseExtractedEvents } from "../llm/response-parser";
import { formatTaxonomyForPrompt } from "../data/taxonomy";
import { encode, embeddingToBuffer } from "../embeddings/bge-m3";
import { insertRawEvents, type RawEventRow } from "../storage/database";
import type { NewsArticle } from "../data/news-fetcher";

const BATCH_SIZE = 3;

export async function extractEventsForDay(
  client: GeminiClient,
  articles: NewsArticle[],
  date: string
): Promise<RawEventRow[]> {
  if (!articles.length) return [];

  const { groups, typeList } = formatTaxonomyForPrompt();
  const allEvents: RawEventRow[] = [];

  for (let i = 0; i < articles.length; i += BATCH_SIZE) {
    const batch = articles.slice(i, i + BATCH_SIZE);

    let articlesText = "";
    for (const [idx, art] of batch.entries()) {
      const body = art.body || art.title;
      articlesText += `\n--- Article ${idx + 1} (source: ${art.source}) ---\n`;
      articlesText += `Title: ${art.title}\n`;
      if (body !== art.title) articlesText += `Content: ${body.slice(0, 2000)}\n`;
    }

    const prompt = fillTemplate(EXTRACT_PROMPT, { groups, type_list: typeList, articles: articlesText });

    try {
      const result = await client.generateJson(prompt);
      const arr = Array.isArray(result) ? result : [result];
      const events = parseExtractedEvents(arr);

      for (const ev of events) {
        allEvents.push({
          news_id: null,
          date,
          asset: batch[0].asset || "ALL",
          event_group: ev.event_group,
          event_type: ev.event_type,
          time: ev.time,
          location: ev.location,
          entities: ev.entities,
          industries: ev.industries,
          description: ev.description,
          extended_attrs: ev.extended_attrs,
        });
      }
    } catch (e: any) {
      console.warn(`Extraction failed for batch at ${i}: ${e.message}`);
    }
  }

  // Compute embeddings
  if (allEvents.length > 0) {
    const descriptions = allEvents.map((e) => e.description);
    const embeddings = await encode(descriptions);
    for (let i = 0; i < allEvents.length; i++) {
      allEvents[i].embedding = embeddingToBuffer(embeddings[i]);
    }
  }

  console.log(`Extracted ${allEvents.length} events for ${date}`);
  return allEvents;
}

export async function runExtraction(
  client: GeminiClient,
  newsByDate: Record<string, NewsArticle[]>
): Promise<Record<string, number[]>> {
  const result: Record<string, number[]> = {};
  for (const date of Object.keys(newsByDate).sort()) {
    const events = await extractEventsForDay(client, newsByDate[date], date);
    if (events.length) {
      const ids = insertRawEvents(events);
      result[date] = ids;
    } else {
      result[date] = [];
    }
  }
  return result;
}

File: bitcoin_stockmem/pipeline/step1_extract.py
"""Step 1: Event Extraction (LLM_ext).

Extracts structured events from raw news articles using Gemini.
Batches multiple short articles per LLM call to reduce API usage.
"""

from __future__ import annotations

import json
import logging
from datetime import datetime

from data.taxonomy import format_taxonomy_for_prompt
from embeddings.bge_m3 import encode, embedding_to_bytes
from llm.gemini_client import GeminiClient
from llm.prompts import EXTRACT_PROMPT
from llm.response_parser import parse_extracted_events
from storage.database import insert_raw_events

logger = logging.getLogger(__name__)

BATCH_SIZE = 3  # articles per LLM call


def extract_events_for_day(
    client: GeminiClient,
    articles: list[dict],
    date: str,
) -> list[dict]:
    """Extract events from a day's news articles.

    Args:
        client: Gemini client instance.
        articles: List of article dicts with 'title', 'body', 'url', etc.
        date: The date string (YYYY-MM-DD).

    Returns:
        List of raw event dicts (ready for DB insertion).
    """
    if not articles:
        return []

    groups_str, type_list_str = format_taxonomy_for_prompt()
    all_events = []

    # Process in batches
    for i in range(0, len(articles), BATCH_SIZE):
        batch = articles[i:i + BATCH_SIZE]

        # Format articles for prompt
        articles_text = ""
        news_id_map = {}  # track which article produced which events
        for idx, art in enumerate(batch):
            body = art.get("body") or art.get("title", "")
            articles_text += f"\n--- Article {idx + 1} (source: {art.get('source', 'unknown')}) ---\n"
            articles_text += f"Title: {art['title']}\n"
            if body != art["title"]:
                articles_text += f"Content: {body[:2000]}\n"  # truncate long articles
            news_id_map[idx] = art.get("news_id")

        prompt = EXTRACT_PROMPT.format(
            groups=groups_str,
            type_list=type_list_str,
            articles=articles_text,
        )

        try:
            result = client.generate_json(prompt)
            if not isinstance(result, list):
                result = [result]
            events = parse_extracted_events(result)
        except Exception as e:
            logger.warning(f"Event extraction failed for batch starting at {i}: {e}")
            continue

        for ev in events:
            all_events.append({
                "news_id": batch[0].get("news_id"),  # approximate mapping
                "date": date,
                "asset": batch[0].get("asset", "ALL"),
                "event_group": ev.event_group,
                "event_type": ev.event_type,
                "time": ev.time,
                "location": ev.location,
                "entities": ev.entities,
                "industries": ev.industries,
                "description": ev.description,
                "extended_attrs": ev.extended_attrs,
            })

    # Compute embeddings
    if all_events:
        descriptions = [e["description"] for e in all_events]
        embeddings = encode(descriptions)
        for ev, emb in zip(all_events, embeddings):
            ev["embedding"] = embedding_to_bytes(emb)

    logger.info(f"Extracted {len(all_events)} events for {date}")
    return all_events


def run_extraction(
    client: GeminiClient,
    news_by_date: dict[str, list[dict]],
) -> dict[str, list[int]]:
    """Run event extraction for multiple dates.

    Args:
        client: Gemini client.
        news_by_date: {date_str: [article_dicts]}.

    Returns:
        {date_str: [raw_event_ids]}.
    """
    result = {}
    for date in sorted(news_by_date.keys()):
        articles = news_by_date[date]
        events = extract_events_for_day(client, articles, date)
        if events:
            ids = insert_raw_events(events)
            result[date] = ids
            logger.info(f"{date}: inserted {len(ids)} raw events")
        else:
            result[date] = []
    return result
